### ISSUE TITLE
Content Negotiation types should be listed in OpenAPI

### ISSUE THREAD

**Issue by Jefftree** [kubernetes/kubernetes#113028]
### What would you like to be added?

Content Negotiation types are currently not listed in OpenAPI. For example, it's possible to query resources with an Accept header `Accept: application/json;as=Table;g=meta.k8s.io;v=v1` to list as tables, but the information is missing in the OpenAPI.

/sig api-machinery

### Why is this needed?

Provides better clarity of all operations supported by a K8s cluster.

**Comment by Jefftree** [comment]
It's small enough that I don't think a KEP is necessary. If there are a lot of design details, feel free to expand the [OpenAPI V3 KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2896-openapi-v3/README.md)

**Comment by apelisse** [comment]
Note that OpenAPI v3 and OpenAPI v2 both expose the content-types in different ways. Is this issue about v2 or v3?

**Comment by Jefftree** [comment]
Both are issues although I'd imagine we'd want to focus our efforts on V3. They build off the same object in go restful and I think it might be possible to fix both at the same time.

**Comment by lengrongfu** [comment]
@Jefftree Hi，What are the flags as=Table;g=meta.k8s.io;v=v1 in the Accept header used to negotiate with the server?  What do as, g, v mean respectively.

My understanding of Content Negotiation is that the client can tell the server the expected data format by setting accept, such as `accept: application/json` or `accept=application/yaml`.

**Comment by Jefftree** [comment]
@lengrongfu Yes you are correct, but additional parameters can also be passed in to request for a specific group,version,kind (g=group, v=version, as=kind). See https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go#L73-L94 where we convert the output depending on whether a supplement GVK was provided.

**Comment by sftim** [comment]
I wonder if we want to list every combination of parameters for these media types - that could be a lot of cardinality.

**Comment by apelisse** [comment]
We should definitely list the types associated with each content-type though, since our current change for API Discovery has entirely different types based on the negotiated content-type.

**Comment by lmktfy** [comment]
We made a slip, you could argue.
Instead of `application/json;as=Table;g=meta.k8s.io;v=v1` it should have been `application/json;q=1;as=Table;g=meta.k8s.io;v=v1`. Notice the extra `q=1` there.

Why? Because `q=1` is (was) a separator. What comes before it is a media type and its parameters; what comes after it is a different set of parameters that apply at the HTTP level, but that don't influence the media type you receive.

---

Here's an example of how that mechanism is supposed to work. Let's say you have something that serves plain text with different line lengths. You might request with:
```http
GET /prose HTTP/1.1
Host: lengthy.prose.example
Connection: close
Accept-Language: en-US
Accept: text/plain;format=flowed;q=0.9, text/plain;format=flowed;q=1;wrapmaxchars=100, text/plain;q=0.5
…
```

…or in English, something like: _I only want plain text. I prefer soft line wrap that my client can reflow, and even then, I prefer the lines to be 100 characters at most._

`format` is part of the definition of `text/plain`; `wrapmaxchars` is part of this specific API. And. `q` is defined by HTTP itself.

---

We could retcon the API server to tolerate / parse / understand the `q` parameter and clients to send `Accept:`.
It's complicated. Using `q` as a separator isn't deprecated, exactly, but having anything after the `q` is. We are definitely sending the after-the-`q` kind of parameter here; JSON (the format) knows nothing of partial object fetches or server-side table formatting.

“Could” doesn't mean “should”. But I think we **shouldn't** add these details to OpenAPI as if to say: you can get a response in any of these formats. Because… they're not media types. They look like they are, but they're not.

Whatever we do about our conformance to HTTP,  I think we gain little by compounding the error. Let's look for another way to declare that you can add (eg) `as=Table;g=meta.k8s.io;v=v1` to an Accept header.

**Comment by lmktfy** [comment]
For watches, I advocate a different route (might need a KEP): use the `Accept:` header to indicate that a client can understand `application/json-seq`, and if if does, send `application/json-seq`.

A watch response is invalid JSON but a valid [JSON text sequence](https://www.rfc-editor.org/rfc/rfc7464). If (really iff) anyone wants to work on this, I think the route is obvious:

* update client-go to accept **watch** responses marked as JSON text sequences
* update kubectl to request JSON text sequences when doing a **watch**
* update kube-apiserver to send JSON text sequence watch responses when these are negotiated
  * and, yes, update the OpenAPI
* update the API documentation
* update the control plane to request JSON text sequence in watches
* …and the kubelet
* eventually, deprecate doing a watch without specifying that you accept JSON text sequences
  * but we **never** remove the legacy support; it stays there as a stable feature of the API

**Comment by lmktfy** [comment]
We could add support for sending CBOR sequences too, for even more efficient watches. I'd like that.

**Comment by seans3** [comment]
/triage accepted
/assign @Jefftree 

Hi Jeff --but movement here? Should this remain open?
